docs/distribution/launcher.md

```diff
@@ -0,0 +1,188 @@
+  - [Downloading GraalVM Releases](#downloading-graalvm-releases)
+- [Running Enso Components](#running-enso-components)
+- [Global User Configuration](#global-user-configuration)
+- [Updating the Launcher](#updating-the-launcher)
+  - [Minimal Required Launcher Version](#minimal-required-launcher-version)
+  - [Downloading Launcher Releases](#downloading-launcher-releases)
+
+<!-- /MarkdownTOC -->
+
+## Launcher Distribution
+The launcher is distributed as a native binary for each platform (Windows,
+Linux, Mac). It is distributed in a ZIP archive as described in
+[Enso Home Layout](./distribution.md#enso-home-layout), except that the
+component directories are empty (they will be populated when Enso versions are
+downloaded by the launcher on-demand). The only non-empty directory is `bin`
+which is where the launcher binary is placed. The ZIP file should also contain a
+README describing the distribution and basic usage.
+
+The distribution structure is portable - this directory structure can be placed
+anywhere in the system. The launcher places all components in this directory
+structure by accessing them as relative to the location of its binary.
+
+It is a good idea to add the `bin` folder to your system `PATH` so that you can
+run the launcher just by typing `enso`.
+
+## Project Management
+The launcher provides basic project management utilities for the command-line
+user.
+
+### Creating a Project
+It allows to create an empty project in a specified directory with the basic
+configuration based on user's config.
+
+### Per-project Enso Version
+Project configuration can specify the exact Enso version that should be used
+inside that project. The launcher automatically detects if it is in a project
+(by traversing the directory structure). The current project can also be
+specified by the `--path` parameter. All components launched inside a project
+use the version specified in the configuration.
+
+> The actionables for this section are:
+> 
+> - Decide how to support inexact project bounds (like `>=3.1, <4` - when should
+>   the launcher check for new versions) and resolvers.
+
+### Project Configuration
+The command-line allows to edit project configuration, for example: change the
+author's name or Enso version.
+
+## Enso and Graal Version Management
+The launcher automatically manages required Enso versions. When running inside a
+project, if the version specified in project configuration is not installed, it
+is installed automatically.
+
+If an Enso version is no longer needed, it can be removed with the `uninstall`
+command.
+
+Moreover, GraalVM distributions tied to the installed Enso versions are managed
+automatically by the launcher. When a new Enso version is installed, it also
+ensures that the correct GraalVM version is installed and that it is used for
+launching this version of Enso. When a managed GraalVM distribution is no longer
+used by any installed version of Enso, it is automatically removed.
```

Comment:
Do we potentially want an `enso cleanup` command that cleans up graal versions not referenced by any installed enso version?

Reply:
My idea was to *automatically* check if the graal version tied to an Enso version that is being uninstalled is used by any other version and if not, to remove it automatically when it is uninstalled. Thus graal versions not referenced by any Enso version should never be present. Unless the user doesn't use our `uninstall` command but removes it manually, but if they manually remove Enso, they might as well remove Graal manually too.

Reply:
Okay, that's fair!